site/src/sphinx/advanced-structured-logging.rst

```diff
@@ -148,6 +152,106 @@ automatically for you:
+    sb.contentPreview(100);
+    ...
+    VirtualHostBuilder vhb = new VirtualHostBuilder("http://example.com");
+    // In this case, the property of virtual host takes precedence over that of server.
+    vhb.contentPreview(150);
+    ...
+    sb.virtualHost(vhb.build());
+
+.. code-block:: java
+
+    import com.linecorp.armeria.client.HttpClientBuilder;
+
+    HttpClientBuilder cb = new HttpClientBuilder();
+    ...
+    cb.contentPreview(100);
+
+Note that the ``contentPreview()`` method enables the previews only for textual content
+which meets one of the following cases:
+
+- when its type matches ``text/*`` or ``application/x-www-form-urlencoded``.
+- when its charset has been specified. e.g. application/json; charset=utf-8.
+- when its subtype is ``xml`` or ``json``. e.g. application/xml, application/json.
+- when its subtype ends with ``+xml`` or ``+json``. e.g. application/atom+xml, application/hal+json
+
+You can also customize the previews by specifying your own :api:`ContentPreviewerFactory` implementation.
+The following example enables the textual preview of first 100 characters for the content type of ``text/*``,
+and the hex dump preview of first 100 bytes for other types:
+
+.. code-block:: java
+
+    import io.netty.buffer.ByteBufUtil;
+    import com.linecorp.armeria.common.MediaType;
+    import com.linecorp.armeria.common.logging.ContentPreviewer;
+
+    ServerBuilder sb = new ServerBuilder();
+
+    sb.contentPreviewerFactory((ctx, headers) -> {
+        MediaType contentType = headers.contentType();
+        if (contentType != null && contentType.is(MediaType.ANY_TEXT_TYPE)) {
+            // Produces the textual preview of the first 100 characters.
+            return ContentPreviewer.ofText(100);
+        }
+        // Produces the hex dump of the first 100 bytes.
+        return ContentPreviewer.ofBinary(100, byteBuf -> {
+            // byteBuf has no more than 100 bytes.
+            return ByteBufUtil.hexDump(byteBuf);
+        });
+    });
+
+You can write your own :api:`ContentPreviewer` to change the way to make the preview. e.g.
+
+.. code-block:: java
+
+    class HexDumpContentPreviewer implements ContentPreviewer {
+        private final StringBuilder builder = new StringBuilder();
+        @Override
+        public void onHeaders(HttpHeaders headers) {
+            // Invoked when headers of a request or response is received.
+        }
+
+        @Override
+        public void onData(HttpData data) {
+            // Invoked when a new content is received.
+            builder.append(ByteBufUtil.hexDump(data.array(), data.offset(), data.length()));
+        }
+
+        @Override
+        public void isDone() {
+            // If it returns true, no further event is called but produce().
+            return false;
+        }
+
+        @Override
+        public String produce() {
+            // Invoked when a request or response ends.
+            return builder.build();
```

Comment:
How about adding some assertion that this method will never be called twice? e.g.

```java

public boolean isDone() {
    return preview != null;
}

public String produce() {
    assert preview == null;
    preview = builder.build();
    return preview;
```